activerecord/lib/active_record/connection_adapters/abstract/query_cache.rb

```diff
@@ -87,6 +87,14 @@ def uncached
       # the same SQL query and repeatedly return the same result each time, silently
       # undermining the randomness you were expecting.
       def clear_query_cache
+        ActiveRecord::Base.connection_handlers.each_value do |handler|
+          handler.connection_pool_list.each do |pool|
+            pool.connections.each(&:clear_query_cache!)
```

Comment:
I don't think we can can iterate over `pool.connections` here without holding the pool's mutex.

Regardless, we don't need to clear the query cache on all connections; we only need read-your-writes consistency for the current thread. If it's acceptable for a concurrent request in another process to see stale data, it should be acceptable for a concurrent request in another thread, too.

```suggestion
            pool.connection.clear_query_cache! if pool.active_connection?
```

Reply:
:+1:

Beyond it being acceptable, having one thread's wipe out every cache in the system would be not-great for hit rates / performance.

Along that line, I think I'd actually go further, and limit this to only the current-thread-owned connection _in the peer pools_ -- the corresponding same-named pool in other handlers -- not all pools.

Reply:
Yea this clearly blows up in the tests but also @eugeneius version doesn't pass either.

I'm starting to get really frustrated with the query cache. It seems like it just wasn't built for multiple connections and we may need to rethink it sooner rather than later.

Reply:
Ok I found a fix I think and I'm not proud of it but the tests pass locally. I'm not sure what consequences to test for at this point if the tests pass on CI.